lib/tests/streamlit/caching_test.py

```diff
@@ -383,13 +385,203 @@ def off_test_has_changes(self):
 
             self.assertEqual(c.value, val)
 
-    @patch.object(st, "warning")
-    def off_test_mutate(self, warning):
+    @patch.object(st, "exception")
+    def off_test_mutate(self, exception):
         for _ in range(2):
             c = st.Cache()
             if c:
                 c.value = [0, 1]
 
             c.value[0] = 1
 
-        warning.assert_called()
+        exception.assert_called()
+
+
+class CacheErrorsTest(testutil.DeltaGeneratorTestCase):
+    """Make sure user-visible error messages look correct.
+
+    These errors are a little annoying to test, but they're important! So we
+    are testing them word-for-word as much as possible. Even though this
+    *feels* like an antipattern, it isn't: we're making sure the codepaths
+    that pull useful debug info from the code are working.
+    """
+
+    def test_st_warning_text(self):
+        @st.cache
+        def st_warning_text_func():
+            st.markdown("hi")
```

Comment:
If you call `st_warning_text_func()` a second time we still see `hi` printed to the screen. 

That sounds like a bug though right, since there should be a cache hit and we won't re-execute the `st.markdown` call? 

Looks like the behavior is already happening on `develop` and wasn't introduced by your code.

Reply:
I don't think that's what's happening when you run the same cached function twice. Try running this, for example:

```python
import streamlit as st

@st.cache
def func():
    st.write("hi")

"first call"
func()

"second call"
func()
```

You'll see this:
```
first call
[warning]
hi
second call
```

So the "hi" belongs to the first call, and there's no "hi" for the second call. That makes sense, since the second call is a cache hit.

Reply:
Ah, I didn't explain myself properly. 

I meant, using your example, if you re-run the report with a single call to `func` then `hi` remains displayed on the screen. 

On the second run of the report, the call to `func` should be a cache hit and `st.write` shouldn't be called, right? That `hi` element shouldn't get sent to the client on re-run.

Reply:
Oh, I think this is the caching bug Tim found. He's fixing that.